spring-data-microservice-transaction-sample/README.md

```diff
@@ -0,0 +1,548 @@
+
+```java
+// Put the order info into the orders table
+orderRepository.insert(order);
+
+AtomicInteger amount = new AtomicInteger();
+for (ItemOrder itemOrder : request.getItemOrderList()) {
+  int itemId = itemOrder.getItemId();
+  int count = itemOrder.getCount();
+  // Retrieve the item info from the items table
+  Optional<Item> itemOpt = itemRepository.findById(itemId);
+  if (!itemOpt.isPresent()) {
+    String message = "Item not found: " + itemId;
+    responseObserver.onError(
+        Status.NOT_FOUND.withDescription(message).asRuntimeException());
+    throw new ScalarDbNonTransientException(message);
+  }
+  Item item = itemOpt.get();
+
+  int cost = item.price * count;
+  // Put the order statement into the statements table
+  statementRepository.insert(new Statement(itemId, orderId, count));
+  // Calculate the total amount
+  amount.addAndGet(cost);
+}
+```
+
+And, Order Service calls the `payment` gRPC endpoint of Customer Service along with the transaction ID (the code is [here](order-service/src/main/java/sample/order/OrderService.java#L131)).
+
+This endpoint first joins the transaction with [join()](customer-service/src/main/java/sample/customer/CustomerService.java#L183):
+
+```java
+if (isJoin) {
+  // Join the transaction
+  customerRepository.join(txId);
+} else {
+  // Resume the transaction
+  customerRepository.resume(txId);
+}
+```
+
+`join()` is called via [execTwoPcOperation()](customer-service/src/main/java/sample/customer/CustomerService.java#L99):
+
+```java
+public void payment(PaymentRequest request, StreamObserver<Empty> responseObserver) {
+  execTwoPcOperation(request.getTransactionId(), true, responseObserver, "Payment", () -> {
+```
+
+It then gets the customer information, and checks if the customer's credit total exceeds the credit limit after the payment.
+And if the check is okay, it updates the customer's credit total (the code is [here](customer-service/src/main/java/sample/customer/CustomerService.java#L100-L114)):
+
+```java
+Customer customer = getCustomer(responseObserver, request.getCustomerId());
+
+int updatedCreditTotal = customer.creditTotal + request.getAmount();
+// Check if the credit total exceeds the credit limit after payment
+if (updatedCreditTotal > customer.creditLimit) {
+  String message = String.format(
+      "Credit limit exceeded. creditTotal:%d, payment:%d", customer.creditTotal, request.getAmount());
+  responseObserver.onError(
+      Status.FAILED_PRECONDITION.withDescription(message).asRuntimeException());
+  throw new ScalarDbNonTransientException(message);
+}
+
+// Reduce credit_total for the customer
+customerRepository.update(customer.withCreditTotal(updatedCreditTotal));
+```
+
+### 3. Two-phase Commit
+
+Once Order Service receives a response that the payment succeeded, Order Service tries to commit the transaction.
```

Comment:
```suggestion
Once the Order Service receives a response that the payment succeeded, the Order Service tries to commit the transaction.
```

Reply:
Applying this change in https://github.com/scalar-labs/scalardb-samples/pull/40